src/charm.py

```diff
@@ -120,15 +142,33 @@ def _login_ui_layer(self) -> Layer:
         return Layer(pebble_layer)
 
     def _get_kratos_endpoint_info(self) -> Optional[str]:
-        kratos_public_url = ""
         if self.model.relations[self._kratos_relation_name]:
             try:
                 kratos_endpoints = self.kratos_endpoints.get_kratos_endpoints()
                 kratos_public_url = kratos_endpoints["public_endpoint"]
+                return kratos_public_url
             except KratosEndpointsRelationDataMissingError:
                 logger.info("No kratos-endpoint-info relation data found")
+                return ""
+        return ""
+
+    def _update_login_ui_endpoint_relation_data(self, event: RelationEvent) -> None:
+        endpoint = self.ingress.url if self.ingress.is_ready() else ""
```

Comment:
It's probably better to return if ingress is not ready and let the other side handle that.

Reply:
Well currently if we do that, than the other components aren't notified if the ingress goes down, or is unavailable. This whole login_endpoints.py get_login_endpoints  method raises an exception if default_url is an empty string fiasco exists, so that the other components aren't left in the dark about changes in proxy state.

That being said, I'm not married to the current approach, but we should probably get together sometime to finalize the design decisions about how the components should integrate.

Reply:
Currently if traefik is not ready, the databag will look like this:
```json
                {
                    "consent_url": "/consent",
                    "error_url": "/error",
                    "index_url": "/index",
                    "login_url": "/login",
                    "oidc_error_url": "/oidc_error",
                    "registration_url": "/registration",
                    "default_url": "",
                }
```
Is this what we want? I don't think we want placeholder values in the databag. I think that the values in the databag should be valid and the other side of the relation (kratos/hydra charms) should be responsible for handling it if the UI is not ready to be consumed.

On that note, we don't have any unit tests for the library (which would have caught such issues). Can you either add them in this PR or open an issue to do it in the future.?

Reply:
This is an ugly solution, I'll concede that. What happens is because of the default_url="", the lib throws a datamissing error, which is then handled by the operator consuming the library. Current compromise I can offer for now is having the databag be all empty values, because I have reservations about not sending updates the hydra and kratos about login_ui proxy change. Alternatively we need some messaging mechanisms between the operators.

Reply:
If you send empty values in the databag, hydra/kratos will not raise `LoginUIEndpointsRelationDataMissingError` or `LoginUIEndpointsRelationMissingError` (and thus not use the default urls that are defined on consumer sides of relation). If they receive empty values, the charms will error out due to missing config